server/src/main/java/io/crate/expression/scalar/ArraySetFunction.java

```diff
@@ -111,4 +116,45 @@ private void setElement(List<Object> source, int index, Object value) {
             source.add(value);
         }
     }
+
+    public static class Allocator {
+        private final LinkedHashMap<Reference, LinkedHashMap<Symbol, Symbol>> mappings;
+        private final BiFunction<String, List<Symbol>, Symbol> functionAllocator;
+
+        public Allocator(BiFunction<String, List<Symbol>, Symbol> functionAllocator) {
+            this.functionAllocator = functionAllocator;
+            this.mappings = new LinkedHashMap<>();
+        }
+
+        public void put(Reference reference, Symbol index, Symbol value) {
+            var mapping = mappings.get(reference);
+            if (mapping != null) {
+                mapping.put(index, value);
+            } else {
+                mapping = new LinkedHashMap<>();
+                mapping.put(index, value);
+                mappings.put(reference, mapping);
+            }
+        }
+
+        public LinkedHashMap<Reference, Symbol> allocate() {
+            LinkedHashMap<Reference, Symbol> refToArraySetMap = new LinkedHashMap<>(mappings.size());
```

Comment:
Is this order sensitive? Otherwise we could use regular `HashMap`s here (and above) as they're cheaper.

Reply:
Thank you, I have `test_update_array_by_elements` which shows that the order needs to be preserved but it does not justify all to be a `LinkedHashMap`, I will try to think of other scenarios.

Reply:
Sorry you're right. The original `LinkedHashMap` was introduced to handle https://github.com/crate/crate/pull/12914. And I think the same should apply to array element updates resulting in dynamic columns, for example:

```
cr> create table t (o array(object), p array(object));                                                                                            
insert CREATE OK, 1 row affected  (1.600 sec)
cr> insert into t (o,p) values ([],[]);                                                                                                           
INSERT OK, 1 row affected  (0.045 sec)
cr> update t set o[2] = {a=1}, p[1] = {c=1}, o[1] = {b=1}; -- notice p['c'] before o['b']                                                                                       
UPDATE OK, 1 row affected  (0.173 sec)

cr> select column_name, ordinal_position from information_schema.columns where table_name = 't' order by 2;                                       
+-------------+------------------+
| column_name | ordinal_position |
+-------------+------------------+
| o           |                1 |
| p           |                2 |
| o['a']      |                3 |
| o['b']      |                4 |  <--- should be 5
| p['c']      |                5 |  <--- should be 4
+-------------+------------------+
SELECT 5 rows in set (0.003 sec)
```

The current implementation cannot handle this scenario, I will take some more time on this. Thanks.

Reply:
I think preserving the ordering across object types is not feasible in current CrateDB implementation. Also, since array element updates will not affect the ordering of the top-level columns (where top-level column ordering affects the ordering of `select *`), I think we can consider this as a limitation or negligible. I will proceed with `LinkedHashMap` > `HashMap` as suggested.